### PR TITLE
qa/suites/fs: mon_thrash test for fs

### PR DESCRIPTION
**Pull request by joscollin**
Created a mon.yaml in fs suite that calls mon_thrash test for fs and multimds suites.

Signed-off-by: Jos Collin <jcollin@redhat.com>

### CODE REVIEW

**Comment by joscollin** [comment]
@batrick The testing is going on. Pushed due to the PC Refreshing.

File: qa/tasks/cephfs/filesystem.py [hunk 1]
```diff
@@ -140,6 +140,31 @@ def get_mds_addr(self, name):
...
+        return False
+
+    def cmpstatus(self, status):
+        """
+        Compares two statuses for mds failovers.
+        Returns True if the statuses are matching.
+        """
+        for fs in status.map['filesystems']:
+            for info in fs['mdsmap']['info'].values():
+                if not self.cmpincname(fs['mdsmap']['fs_name'], info['incarnation']):
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py]: these parameters don't match the parameter names for `cmpincname`.

File: qa/tasks/mon_thrash.py [hunk 2]
```diff
@@ -212,8 +214,12 @@ def max_killable(self):
 
     def do_thrash(self):
         """
-        Cotinuously loop and thrash the monitors.
+        Continuously loop and thrash the monitors.
         """
+        #status before mon thrashing
+        beforestatus = self.mds_cluster.status()
+        fs = beforestatus.get_filesystems()
```
Comment (batrick) [inline qa/tasks/mon_thrash.py]: We should check that every file system did not experience a failover.

Comment (batrick) [reply qa/tasks/mon_thrash.py]: Also, we should only do these checks when enabled for cephfs thrash testing. i.e. make this check a config option to `mon_thrash.py`.

File: qa/tasks/mon_thrash.py [hunk 3]
```diff
@@ -307,6 +313,27 @@ def do_thrash(self):
                     delay=self.thrash_delay))
                 time.sleep(self.thrash_delay)
 
+        #status after thrashing
+        status = self.mds_cluster.status()
+
+        # make sure everyone is in active, standby, or standby-replay
+        log.info('Wait for all MDSs to reach steady state...')
```
Comment (batrick) [inline qa/tasks/mon_thrash.py]: I don't see a reason to wait for all MDSs to reach a stable state? We just need to check if any MDSs have failed.

File: qa/tasks/mon_thrash.py (hunk 2 above)
Comment (joscollin) [reply qa/tasks/mon_thrash.py]: @batrick : 

> We should check that every file system did not experience a failover.

Here we are comparing two statuses. Each filesystems are considered inside `hadfailover` function. Are you suggesting to bring that filesystem iteration outside `class FSStatus`? Please clarify.

> make this check a config option to mon_thrash.py
I will make it config option.

File: qa/tasks/cephfs/filesystem.py [hunk 4]
```diff
@@ -140,6 +140,31 @@ def get_mds_addr(self, name):
...
+        return False
+
+    def hadfailover(self, status):
+        """
+        Compares two statuses for mds failovers.
+        Returns True if the statuses are matching.
+        """
+        for fs in status.map['filesystems']:
+            for info in fs['mdsmap']['info'].values():
+                if not self.cmpincname(info['incarnation'], fs['mdsmap']['fs_name']):
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py]: At this point you have the MDS info which includes a unique `gid`. Let's use that so you can look it up.

````
def get_mds_gid(self, gid):
    # like get_mds() but with gid
````

Then you can do in this code:

````
oldinfo = self.get_mds_gid(info['gid'])
if oldinfo is None or oldinfo['incarnation'] != info['incarnation']:
...
````

and then drop the `cmpincname` method.

File: qa/suites/fs/thrash/ceph-thrash/mon.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+overrides:
+  ceph:
+    log-whitelist:
+      - overall HEALTH_
+      - \(MON_DOWN\)
+tasks:
+- mon_thrash:
```
Comment (batrick) [inline qa/suites/fs/thrash/ceph-thrash/mon.yaml]: set `check_mds_failover` here?

File: qa/tasks/mon_thrash.py [hunk 6]
```diff
@@ -59,6 +60,7 @@ class MonitorThrasher:
                         in % (default: 0)
     freeze_mon_duration: how many seconds to freeze the mon (default: 15)
     scrub               Scrub after each iteration (default: True)
+    mds_failover        Check if mds failover happened (default: False)
```
Comment (batrick) [inline qa/tasks/mon_thrash.py]: Let's rename `check_mds_failover`.

File: qa/tasks/cephfs/filesystem.py (hunk 4 above)
Comment (joscollin) [reply qa/tasks/cephfs/filesystem.py]: That's better way of doing it! I saw that code earlier, but didn't thought doing that way. Thanks.

File: qa/suites/fs/thrash/ceph-thrash/mon.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+    check_mds_failover: False
```
Comment (batrick) [inline qa/suites/fs/thrash/ceph-thrash/mon.yaml]: `True`?

File: qa/tasks/cephfs/filesystem.py [hunk 8]
```diff
@@ -140,6 +140,28 @@ def get_mds_addr(self, name):
...
+    def hadfailover(self, status):
+        """
+        Compares two statuses for mds failovers.
+        Returns True if the statuses are matching.
+        """
+        for fs in status.map['filesystems']:
+            for info in fs['mdsmap']['info'].values():
+                oldinfo = self.get_mds_gid(info['gid'])
+                if oldinfo is None or oldinfo['incarnation'] != info['incarnation']
+                    return False
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py]: The returns are reversed. `hadfailover` shoudl return True here.

File: qa/tasks/cephfs/filesystem.py [hunk 9]
```diff
@@ -140,6 +140,28 @@ def get_mds_addr(self, name):
...
+        #all matching
+        return True
```
Comment (batrick) [inline qa/tasks/cephfs/filesystem.py]: `False`

File: qa/tasks/mon_thrash.py [hunk 10]
```diff
@@ -307,6 +320,13 @@ def do_thrash(self):
                     delay=self.thrash_delay))
                 time.sleep(self.thrash_delay)
 
+        if self.mds_failover:
+            #status after thrashing
+            status = self.mds_cluster.status()
+            assert beforestatus.hadfailover(status), \
```
Comment (batrick) [inline qa/tasks/mon_thrash.py]: `assert not`

File: qa/tasks/mon_thrash.py [hunk 11]
```diff
@@ -127,6 +130,12 @@ def __init__(self, ctx, manager, config, logger):
             assert self.maintain_quorum, \
                 'store_thrash = true must imply maintain_quorum = true'
 
+        #MDS failover
+        self.mds_failover = self.config.get('check_mds_failover', True)
```
Comment (batrick) [inline qa/tasks/mon_thrash.py]: default `False`

**Comment by batrick** [comment]
otherwise lgtm

**Pull request merged**